### ISSUE TITLE
 Issue with async_method_call_timed not respecting timeout

### ISSUE THREAD

**Issue by ClaudiusJXK** [openbmc/sdbusplus#98]
Description:

I provide a dbus method call on the Server side that takes a long time, requiring 10 seconds to complete.
On the Client side, I initiate a method call to the Server using the async_method_call_timed interface from connection.hpp, setting a timeout of 1 second.
The callback passed to async_method_call_timed on the Client side is not invoked after 1 second as expected, but rather after the full 10 seconds it takes for the server-side execution to complete.

Server Code:
```cpp
#include <boost/asio/io_context.hpp>
#include <memory>
#include <sdbusplus/asio/object_server.hpp>
#include <iostream>
#include <chrono>
#include <ctime>
#include <iomanip>
#include <string>#include <iomanip>


void outputTime(const std::string& from)
{
    auto now = std::chrono::system_clock::now();
    std::time_t now_time = std::chrono::system_clock::to_time_t(now);
    std::cout << from <<": " << "Current time: " << std::put_time(std::localtime(&now_time), "%Y-%m-%d %H:%M:%S") << std::endl;
}

boost::asio::io_context io;
auto conn = std::make_shared<sdbusplus::asio::connection>(io);
sdbusplus::asio::object_server server(conn);


const auto Service = "xyz.openbmc.test";
const auto Path = "/xyz/openbmc/test";
const auto Interface = "xyz.openbmc.test.interface";


void registService()
{
    conn->request_name(Service);
}

void outputTime()
{
    auto now = std::chrono::system_clock::now();
    // 将时间点转换为时间_t
    std::time_t now_time = std::chrono::system_clock::to_time_t(now);
    // 将 time_t 转换为可读格式
    std::cout << "Current time: " << std::put_time(std::localtime(&now_time), "%Y-%m-%d %H:%M:%S") << std::endl;
}

void pulicInterface()
{
    auto intf = server.add_interface(Path, Interface);
    intf->register_method("Method", [](std::string& input){
            std::cout << "server: method" << "receive request" << std::endl;
            outputTime("server");
            if( input == "Sleep" )
            {
                sleep(10);
            }
            std::cout << "server: " << "sleep end" << std::endl;
            return 1;
    });
    intf->initialize();
}



int main(int, char**)
{
    registService();
    pulicInterface();
    io.run();
    return 0;
}
```

Client Code:
```cpp
#include <boost/asio/io_context.hpp>
#include <sdbusplus/asio/object_server.hpp>
#include "sdbusplus/asio/connection.hpp"
#include <iostream>
#include <chrono>
#include <ctime>
#include <iomanip>
#include <string>
#include <iomanip>


void outputTime(const std::string& from)
{
    auto now = std::chrono::system_clock::now();
    std::time_t now_time = std::chrono::system_clock::to_time_t(now);
    std::cout << from <<": " << "Current time: " << std::put_time(std::localtime(&now_time), "%Y-%m-%d %H:%M:%S") << std::endl;
}

const auto BASE_INTF = "org.freedesktop.DBus.Properties";
const auto METHOD_GET = "Get";
const auto METHOD_SET = "Set";
const auto METHOD_GETALL = "GetAll";

const auto Service = "xyz.openbmc.test";
const auto Path = "/xyz/openbmc/test";
const auto Interface = "xyz.openbmc.test.interface";


boost::asio::io_context io;
auto conn = std::make_shared<sdbusplus::asio::connection>(io);

void call()
{
    std::cout << "before called: " << std::endl;
    outputTime("client");

    conn->async_method_call_timed(
        [](boost::system::error_code ec)
        {
            if (ec)
            {
                std::cout << "error happend: " << ec.message() << std::endl;
                outputTime("client");
                return;
            }

            std::cout << "every thing is ok "  << std::endl;
            outputTime("client");
        },
        Service,
        Path,
        Interface,
        "Method",
        1*1000*1000,
        "Sleep"
    );
}


int main(int , char** )
{
    call();
    io.run();

    return 0;
}
```
callback in client  will invoke in 10s later, and report timeout error. but timeout param  I passed  was 1s.
when I change io.run() in client to conn->process_loop();  callback will be invoke in 1s later, that was i expected.

**Comment by williamspatrick** [comment]
@edtanous Is this something you want to look at?

**Comment by ClaudiusJXK** [comment]
Does the sd_bus file descriptor remain unreadable even after a timeout, causing it not to be scheduled by boost::asio? @williamspatrick @edtanous

**Comment by edtanous** [comment]
This isn't something I'm able to look at this moment, but I can add it to my list.  i'm not clear to me how libsystemd implements its timeouts internally, so we'll have to identify if the timeout parameter is getting to libsystemd correctly, and then what behavior it's expecting us to handle.  It's possible we're only waiting on read events from the descriptor, and we also need to wait on other events:

The first thing I'd start to do is to move this call:
https://github.com/openbmc/sdbusplus/blob/c012fde41552c478d45145b7c01db88df1070ff2/include/sdbusplus/asio/connection.hpp#L317

to the more modern socket.async_wait() with a wait_type of wait_read and wait_error.  null_buffers() was technically correct at the time we wrote that code, but it's been superseded by more specific implementations.

if someone wants to try that before I get there and report back, feel free.

**Comment by edtanous** [comment]
What I assume is the fix turned out to not be as hard as I thought it would be.

@ClaudiusJXK Could you please retest your test applications against the commit below, and let me know if it solves the problem?

http://gerrit.openbmc.org/c/openbmc/sdbusplus/+/75251

**Comment by ClaudiusJXK** [comment]
> This isn't something I'm able to look at this moment, but I can add it to my list. i'm not clear to me how libsystemd implements its timeouts internally, so we'll have to identify if the timeout parameter is getting to libsystemd correctly, and then what behavior it's expecting us to handle. It's possible we're only waiting on read events from the descriptor, and we also need to wait on other events:我现在无法查看这一点，但我可以将其添加到我的列表中。我不清楚 libsystemd 如何在内部实现其超时，因此我们必须确定 timeout 参数是否正确到达 libsystemd，然后确定它希望我们处理什么行为。我们可能只等待来自描述符的 read 事件，我们还需要等待其他事件：
> 
> The first thing I'd start to do is to move this call:我要开始做的第一件事是移动此调用：
> 
> https://github.com/openbmc/sdbusplus/blob/c012fde41552c478d45145b7c01db88df1070ff2/include/sdbusplus/asio/connection.hpp#L317
> 
> to the more modern socket.async_wait() with a wait_type of wait_read and wait_error. null_buffers() was technically correct at the time we wrote that code, but it's been superseded by more specific implementations.转换为具有 wait_read 和 wait_error wait_type的更现代的 socket.async_wait（）。在我们编写该代码时，null_buffers（） 在技术上是正确的，但它已被更具体的实现所取代。
> 
> if someone wants to try that before I get there and report back, feel free.如果有人想在我到达那里并报告之前尝试一下，请随时。

**Comment by ClaudiusJXK** [comment]
> This isn't something I'm able to look at this moment, but I can add it to my list. i'm not clear to me how libsystemd implements its timeouts internally, so we'll have to identify if the timeout parameter is getting to libsystemd correctly, and then what behavior it's expecting us to handle. It's possible we're only waiting on read events from the descriptor, and we also need to wait on other events:
> 
> The first thing I'd start to do is to move this call:
> 
> https://github.com/openbmc/sdbusplus/blob/c012fde41552c478d45145b7c01db88df1070ff2/include/sdbusplus/asio/connection.hpp#L317
> 
> to the more modern socket.async_wait() with a wait_type of wait_read and wait_error. null_buffers() was technically correct at the time we wrote that code, but it's been superseded by more specific implementations.
> 
> if someone wants to try that before I get there and report back, feel free.


I am confident that the timeout has been passed to systemd, as the timeout callback works correctly when using conn->process_loop().

**Comment by ClaudiusJXK** [comment]
> This isn't something I'm able to look at this moment, but I can add it to my list. i'm not clear to me how libsystemd implements its timeouts internally, so we'll have to identify if the timeout parameter is getting to libsystemd correctly, and then what behavior it's expecting us to handle. It's possible we're only waiting on read events from the descriptor, and we also need to wait on other events:
> 
> The first thing I'd start to do is to move this call:
> 
> https://github.com/openbmc/sdbusplus/blob/c012fde41552c478d45145b7c01db88df1070ff2/include/sdbusplus/asio/connection.hpp#L317
> 
> to the more modern socket.async_wait() with a wait_type of wait_read and wait_error. null_buffers() was technically correct at the time we wrote that code, but it's been superseded by more specific implementations.
> 
> if someone wants to try that before I get there and report back, feel free.

I am confident that the timeout has been passed to systemd, as the timeout callback works correctly when using conn->process_loop().

**Comment by ClaudiusJXK** [comment]
> What I assume is the fix turned out to not be as hard as I thought it would be.
> 
> @ClaudiusJXK Could you please retest your test applications against the commit below, and let me know if it solves the problem?
> 
> http://gerrit.openbmc.org/c/openbmc/sdbusplus/+/75251

I will try latter. tks!

**Comment by edtanous** [comment]
> I am confident that the timeout has been passed to systemd, as the timeout callback works correctly when using conn->process_loop().

Yes, the timeout was being passed to systemd, but it turns out that system expects a user to then call a "get timeout" function that to their poll loop.   Asio connections weren't doing this, so timeouts weren't being obeyed if there was no other traffic.  I suspect in practice there's not a lot of daemons that go completely idle, so that covered up these timeouts not being applied in a lot of cases.

**Comment by ClaudiusJXK** [comment]
> What I assume is the fix turned out to not be as hard as I thought it would be.
> 
> @ClaudiusJXK Could you please retest your test applications against the commit below, and let me know if it solves the problem?
> 
> http://gerrit.openbmc.org/c/openbmc/sdbusplus/+/75251

Thank you very much, the issue has been verified and resolved. @edtanous
